src/main/drivers/dshot_bitbang.c

```diff
@@ -0,0 +1,724 @@
+        }
+
+
+        bbMotors[motorIndex].enabled = true;
+
+        // Fill in motors structure for 4way access (XXX Should be refactored)
+
+        motors[motorIndex].enabled = true;
+    }
+}
+
+static motorVTable_t bbVTable = {
+    .postInit = bbPostInit,
+    .enable = bbEnableMotors,
+    .disable = bbDisableMotors,
+    .isMotorEnabled = bbIsMotorEnabled,
+    .updateStart = bbUpdateStart,
+    .write = bbWrite,
+    .writeInt = bbWriteInt,
+    .updateComplete = bbUpdateComplete,
+    .convertExternalToMotor = dshotConvertFromExternal,
+    .convertMotorToExternal = dshotConvertToExternal,
+    .shutdown = bbShutdown,
+};
+
+dshotBitbangStatus_e dshotBitbangGetStatus()
+{
+    return bbStatus;
+}
+
+motorDevice_t *dshotBitbangDevInit(const motorDevConfig_t *motorConfig, uint8_t count)
+{
+    dbgPinInit();
+    dbgPinLo(0);
+    dbgPinLo(1);
+
+    lastMotorConfig = motorConfig;
+    bbDevice.vTable = bbVTable;
+    motorCount = count;
+    bbStatus = DSHOT_BITBANG_OK;
+
+#ifdef USE_DSHOT_TELEMETRY
+    useDshotTelemetry = motorConfig->useDshotTelemetry;
+#endif
+
+    for (int motorIndex = 0; motorIndex < MAX_SUPPORTED_MOTORS && motorIndex < motorCount; motorIndex++) {
+        const timerHardware_t *timerHardware = timerGetByTag(motorConfig->ioTags[motorIndex]);
+        const IO_t io = IOGetByTag(motorConfig->ioTags[motorIndex]);
+
+        uint8_t output = motorConfig->motorPwmInversion ?  timerHardware->output ^ TIMER_OUTPUT_INVERTED : timerHardware->output;
+        bbPuPdMode = (output & TIMER_OUTPUT_INVERTED) ? BB_GPIO_PULLDOWN : BB_GPIO_PULLUP;
+
+#ifdef USE_DSHOT_TELEMETRY
+        if (useDshotTelemetry) {
+            output ^= TIMER_OUTPUT_INVERTED;
+        }
+#endif
+
+        if (!IOIsFreeOrPreinit(io)) {
+            /* not enough motors initialised for the mixer or a break in the motors */
+            bbDevice.vTable.write = motorWriteNull;
+            bbDevice.vTable.updateComplete = motorUpdateCompleteNull;
+            bbStatus = DSHOT_BITBANG_MOTOR_PIN_CONFLICT;
+            return NULL;
+        }
+
+        int pinIndex = IO_GPIOPinIdx(io);
+
+        bbMotors[motorIndex].pinIndex = pinIndex;
+        bbMotors[motorIndex].io = io;
+        bbMotors[motorIndex].output = output;
+#if defined(STM32F4) || defined(STM32F3)
+        bbMotors[motorIndex].iocfg = IO_CONFIG(GPIO_Mode_OUT, GPIO_Speed_50MHz, GPIO_OType_PP, bbPuPdMode);
+#elif defined(STM32F7)
+        bbMotors[motorIndex].iocfg = IO_CONFIG(GPIO_MODE_OUTPUT_PP, GPIO_SPEED_FREQ_VERY_HIGH, bbPuPdMode);
+#endif
+
+        IOInit(io, OWNER_MOTOR, RESOURCE_INDEX(motorIndex));
+        IOConfigGPIO(io, bbMotors[motorIndex].iocfg);
+        if (output & TIMER_OUTPUT_INVERTED) {
+            IOLo(io);
+        } else {
+            IOHi(io);
+        }
+
+        // Fill in motors structure for 4way access (XXX Should be refactored)
+        motors[motorIndex].io = bbMotors[motorIndex].io;
+    }
+
+    return &bbDevice;
+}
+
+typedef struct dbgPin_s {
```

Comment:
I am still intrigued by what this does? @jflyper?
Looks like some sort of debug facility, and in this case it should probably be extracted into separate files, and put into a separate pull request.

Reply:
Yes, it's a low latency timing marker facility mainly used for precise timing measurements. (Pinio can serve for casual measurements.) Lemme see how I can generalize this and make them into an independent facility.